clang/include/clang/3C/DeclRewriter.h

```diff
@@ -0,0 +1,128 @@
+  void rewrite(RSet &ToRewrite);
+
+  // Rewrite a specific variable declaration using the replacement string in the
+  // DAndReplace structure. Each of these functions is specialized to handling
+  // one subclass of declarations.
+  void rewriteParmVarDecl(ParmVarDeclReplacement *N);
+
+  template<typename DRType>
+  void rewriteFieldOrVarDecl(DRType *N, RSet &ToRewrite);
+  void rewriteMultiDecl(DeclReplacement *N, RSet &ToRewrite);
+  void rewriteSingleDecl(DeclReplacement *N, RSet &ToRewrite);
+  void doDeclRewrite(SourceRange &SR, DeclReplacement *N);
+
+  void rewriteFunctionDecl(FunctionDeclReplacement *N);
+  void getDeclsOnSameLine(DeclReplacement *N, std::vector<Decl *> &Decls);
+  bool isSingleDeclaration(DeclReplacement *N);
+  bool areDeclarationsOnSameLine(DeclReplacement *N1, DeclReplacement *N2);
+  SourceRange getNextCommaOrSemicolon(SourceLocation L);
+};
+
+// Visits function declarations and adds entries with their new rewritten
+// declaration to the RSet RewriteThese.
+class FunctionDeclBuilder : public RecursiveASTVisitor<FunctionDeclBuilder> {
+public:
+  explicit FunctionDeclBuilder(ASTContext *C, ProgramInfo &I, RSet &DR,
+                               std::map<std::string, std::string> &NewFuncSig,
+                               ArrayBoundsRewriter &ArrRewriter)
+      : Context(C), Info(I), RewriteThese(DR), ABRewriter(ArrRewriter),
+        VisitedSet(), ModifiedFuncSignatures(NewFuncSig) {}
+
+  bool VisitFunctionDecl(FunctionDecl *);
+  bool isFunctionVisited(std::string FuncName);
+protected:
+
+  ASTContext            *Context;
+  ProgramInfo           &Info;
+  RSet                  &RewriteThese;
+  ArrayBoundsRewriter   &ABRewriter;
+
+  // Set containing the names of all functions visited in the AST traversal.
+  // Used to ensure the new signature is only computed once for each function.
+  std::set<std::string> VisitedSet;
+
+  // This is a map from functions (the string representation of their names) to
+  // their function signature in the rewritten program.
+  std::map<std::string, std::string> &ModifiedFuncSignatures;
+
+  // Get existing itype string from constraint variables.
+  std::string getExistingIType(ConstraintVariable *DeclC);
+  virtual void buildDeclVar(PVConstraint *Defn, DeclaratorDecl *Decl,
+                            std::string &Type, std::string &IType,
+                            bool &RewriteParm, bool &RewriteRet);
+  void buildCheckedDecl(PVConstraint *Defn, DeclaratorDecl *Decl,
+                        std::string &Type, std::string &IType,
+                        bool &RewriteParm, bool &RewriteRet);
+  void buildItypeDecl(PVConstraint *Defn, DeclaratorDecl *Decl,
+                      std::string &Type, std::string &IType,
+                      bool &RewriteParm, bool &RewriteRet);
+};
+
+class FieldFinder : public RecursiveASTVisitor<FieldFinder> {
+  public:
+    FieldFinder(GlobalVariableGroups &GVG) : GVG(GVG) { }
+
+    bool VisitFieldDecl(FieldDecl *FD);
```

Comment:
I see that some functions use `functionName` while others use `FunctionName` as the naming convention. Could you please make the function cases uniform. Please refer to the LLVM Coding Standards at https://llvm.org/docs/CodingStandards.html

Reply:
It turns out that this occurrence of `VisitFieldDecl` isn't our fault (it's a pseudo-override of `RecursiveASTVisitor::VisitFieldDecl` using the Curiously Recurring Template Pattern), but our code may well have other naming violations, and I'm working on setting up clang-tidy to find them all.

Reply:
I've completed a clang-tidy pass, and this PR is now fully compliant with the implementation of the LLVM naming guideline in the `readability-identifier-naming` rule of clang-tidy version 11.0.0, except for documented suppressions. Please let me know if you have any concerns about our code violating aspects of the guideline that are not enforced by clang-tidy.

To expand on my previous comment: There are several exceptional scenarios in which we can't change names in our code because they have to match names that are not under our control. In some scenarios, clang-tidy doesn't recognize the coupling and would warn about the name, so we have a documented suppression comment. In other scenarios, clang-tidy version 11.0.0 (which we are using and this repository will soon be using!) knows not to warn, though the reason may not always be obvious to a person looking at the code. The scenarios I'm aware of that fall into this second category are:

1. Virtual overriding methods. In the future, we could ensure these are always obvious by enforcing the use of the `override` keyword using the `modernize-use-override`.
2. "Pseudo-overriding" methods that are referenced by name from a template and hide a method of the same name in a superclass that provides a default implementation. Methods named `Visit*` in subclasses of `RecursiveASTVisitor` generally fall into this category.

So if you see names in our code that look bad and don't have a documented suppression, you can either (a) run clang-tidy version 11.0.0, see that it doesn't warn, and trust that there's a reason for that; (b) research yourself which exceptional scenario applies; or (c) ask me. :slightly_smiling_face: 

We are mostly compliant with the other standard clang-tidy rules for the LLVM codebase too.